src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt

```diff
@@ -0,0 +1,43 @@
+package org.wycliffeassociates.otter.common.recorder
+
+import org.wycliffeassociates.otter.common.collections.FloatRingBuffer
+
+class PCMCompressor(private val ringBuffer: FloatRingBuffer) {
+
+    //size is how small the waveform is being shrunk for visualization
+    //arbitrary for now, could be based on number of seconds to show and resolution
+    val accumulator = FloatArray(240)
+    var index = 0
+
+    fun add(data: FloatArray) {
+        for (sample in data) {
+            if (index >= accumulator.size) {
+                sendDataToRingBuffer()
+                index = 0
+            }
+            accumulator[index] = sample
+            index++
+        }
+    }
+
+    fun add(data: Float) {
+        if (index >= accumulator.size) {
+            sendDataToRingBuffer()
+            index = 0
+        }
+        accumulator[index] = data
+        index++
+    }
+
+    fun sendDataToRingBuffer() {
+        var min = Float.MAX_VALUE
+        var max = Float.MIN_VALUE
+
+        for (sample in accumulator) {
+            if (max < sample) max = sample
```

Comment:
I could.
Those return nullable types, so I'd have to handle that.
```
/**
 * Returns the smallest element or `null` if there are no elements.
 * 
 * If any of elements is `NaN` returns `NaN`.
 */
public fun FloatArray.min(): Float? {
    if (isEmpty()) return null
    var min = this[0]
    if (min.isNaN()) return min
    for (i in 1..lastIndex) {
        val e = this[i]
        if (e.isNaN()) return e
        if (min > e) min = e
    }
    return min
}
```

This also means I'd be doing two loops, instead of just one.

Reply:
Interesting points. And yes, you'd have to handle nulls and do two loops, so I see why you'd want to keep it this way.